charts/prometheus-node-exporter/templates/service.yaml

```diff
@@ -5,10 +5,18 @@ metadata:
   namespace: {{ include "prometheus-node-exporter.namespace" . }}
   labels:
     {{- include "prometheus-node-exporter.labels" $ | nindent 4 }}
+  {{- if or $.Values.prometheus.monitor.enabled $.Values.prometheus.podMonitor.enabled }}
   {{- with .Values.service.annotations }}
   annotations:
+    {{- unset . "prometheus.io/scrape" | toYaml | nindent 4 }}
+  {{- end }}
+  {{- else }}
+  annotations:
+    prometheus.io/scrape: "true"
+  {{- with .Values.service.annotations }}
     {{- toYaml . | nindent 4 }}
   {{- end }}
```

Comment:
Is this really necessary?

Reply:
That with block is required since the default value for `service.annotations` is now an empty map which could cause issues if the user does not set any annotations

Reply:
@mhowell24 
I mean, if `service.annotations` is empty, e.g. `{}`, then that will render like this:

```yaml
  annotations:
    prometheus.io/scrape: "true"
    {}                          # <== ???
```

I really don't think that is necessary.

Reply:
We need it for when `.Values.service.annotations` is empty. If we just use the value then when it is empty we try templating an empty map and you end up with invalid yaml/json.

 For example removing the with and updating to this

<img width="475" alt="Screenshot 2023-03-17 at 09 21 17" src="https://user-images.githubusercontent.com/72545593/225864174-60d1e5bf-fcac-4483-a6bf-f90a1a64745d.png">

results in this error when there are no annotations specified

<img width="1065" alt="Screenshot 2023-03-17 at 09 22 18" src="https://user-images.githubusercontent.com/72545593/225864431-dee96618-4b81-4410-80e8-144e70869264.png">


Using the with allows us to only template in the user specific annotations if there are any.

Reply:
@zanhsieh what do you think or is there some other way you would prefer I deal with that?

Reply:
@mhowell24 I mean like this:

```yaml
  ...
  labels:
    {{- include "prometheus-node-exporter.labels" $ | nindent 4 }}
  annotations:
  {{- if or .Values.prometheus.monitor.enabled .Values.prometheus.podMonitor.enabled }}
  {{- with .Values.service.annotations }}
    {{- unset . "prometheus.io/scrape" | toYaml | nindent 4 }}
  {{- end }}
  {{- else }}
    prometheus.io/scrape: "true"
  {{- end }}
  ...
```

Reply:
That stops the user from being able to add additional annotations to the service if they are not using the `podMonitor` or `serviceMonitor`, which is a valid use case. We also need the with block to contain the `annotations` heading so it only gets added if there are annotaions to be added.

I am using the with block on the else branch of that if statement to only render user speficied annotations if there are any. If there are not any specified then we end up passing an empty map to `toYaml` and helm fails to template, as I showed in my above comment. We need with blocks in both branches of the if statement to cover the case of the user specifying and not specifying annotions.